design-proposals/vm-pool.md

```diff
@@ -0,0 +1,194 @@
+* VMConfig and VM object’s spec structure are identical
+
+## VirtualMachinePool (VMPool) API
+
+The VMPool api represents all the tunings necessary for managing a pool of stateful VMs. This API is very simple and leverages the VMConfig object for most of the heavy lifting pertaining to what the VM instances within the pool actually look like.
+
+The VMPools spec contains the following tunings and values
+
+* **VirtualMachineRef** - (Required) A structure that points to a VMConfig object
+* **Replicas** - (Required) An integer representing the desired number of VM replicas
+* **MaxUnavailable**  - (Optional) (Defaults to 25%) Integer or string pointer, that when set represents either a percentage or number of VMs in a pool that can be unavailable (ready condition false) at a time during automated update.
+* **NameGeneration** - (Optional) Specifies how objects within a pool have their names generated
+	* **VMPrefix (Defaults to VMPool.Name)** - String representing the prefix to use for when naming VMs.
+	* **AppendPostfixToSecretReferences: (default false)** - Boolean that indicates if VM’s unique postfix should be appended to references to Secrets in the VMI’s Volumes list. This is useful when needing to pre-generate unique secrets for VMs within a pool.
+	* **AppendPostfixToConfigMapReferences (default false)** - Boolean that indicates if VM’s unique postfix should be appended to ConfigMap references in the VMI’s Volumes list. This is useful when needing to pre-generate unique secrets for VMs within a pool.
+* **DisruptionStrategy** - (Optional) Specifies how the VMPool controller manages scaling in and updating VMs within a VMPool
+	* **Unmanaged** - No automation during scale-in or updates. The VM is never touched after creation. Users manually delete and update individual VMs in a pool. Persistent state preservation is up to the user removing the VMs
+	* **Opportunistic** - Opportunistic scale-in and update of VMs which are in a halted state.
+		* **ScaleInStatePreservation** - (Optional) specifies if and how to preserve state of VMs selected for scale-in.
+			* **Disabled** - (Default) all state for VMs selected for scale-in will be deleted
+			* **Offline** - PVCs for VMs selected for scale-in will be preserved and reused on scale-out (decreases provisioning time during scale out)
+			* **Online** - [NOTE we can't implement this until we have the ability to suspend VM memory state to a PVC] PVCs and memory for VMs selected for scale-in will be preserved and reused on scale-out (decreases provisioning and boot time during scale out)
+Each VM’s PVCs are preserved for future scale out
+	* **Proactive** - Proactive scale-in and update by forcing VMs to shutdown during scale-in and restart during update.
+		* **ScaleInPriority** - The priority in which VM instances are selected for proactive scale-in
+			* **Random** (Default) - Pick VMs by random for scale-in first
+			* **Oldest** - Pick oldest VMs by creation time for scale-in first
```

Comment:
How about newest as well?

Reply:
sure, that could make sense, i picked two here just to show that it's an enum